src/cid.js

```diff
@@ -0,0 +1,284 @@
+        getSourceOrigin(url) +
+        externalCidScope);
+  });
+}
+
+/**
+ * Returns whether the URL has the origin of a proxy
+ * @param {Url} url URL of an AMP document.
+ * @return {boolean}
+ * @visibleForTesting BUT if this is needed elsewhere it could be
+ *     factored into its own package.
+ */
+export function isProxyOrigin(url) {
+  return (url.origin == 'https://cdn.ampproject.org' ||
+      url.origin.indexOf('http://localhost:') == 0);
+}
+
+/**
+ * Returns the source origin of an AMP document for documents served
+ * on a proxy origin. Throws if the doc is not on a proxy origin.
+ * @param {Url} url URL of an AMP document.
+ * @return {string} The source origin of the URL.
+ * @visibleForTesting BUT if this is needed elsewhere it could be
+ *     factored into its own package.
+ */
+export function getSourceOrigin(url) {
+  // List of well known proxy hosts. New proxies must be added here
+  // to generate correct tokens.
+  assert(isProxyOrigin(url), 'Expected proxy origin %s', url.origin);
+  // Example path that is being matched here.
+  // https://cdn.ampproject.org/c/s/www.origin.com/foo/
+  // The /s/ is optional and signals a secure origin.
+  const path = url.pathname.split('/');
+  const prefix = path[1];
+  assert(prefix == 'c' || prefix == 'v',
+      'Unknown path prefix in url %s', url.href);
+  const domainOrHttpsSignal = path[2];
+  const origin = domainOrHttpsSignal == 's'
+      ? 'https://' + path[3]
+      : 'http://' + domainOrHttpsSignal;
+  // Sanity test that what we found looks like a domain.
+  assert(origin.indexOf('.') > 0, 'Expected a . in origin %s Foo', origin);
+  return origin;
+}
+
+/**
+ * Returns the base cid for the current user. This string must not
+ * be exposed to users without hashing with the current source origin
+ * and the externalCidScope.
+ * On a proxy this value is the same for a user across all source
+ * origins.
+ * @param {!Cid} cid
+ * @param {!Promise} persistenceConsent
+ * @return {!Promise}
+ */
+function getBaseCid(cid, persistenceConsent) {
+  if (cid.baseCid_) {
+    return Promise.resolve(cid.baseCid_);
+  }
+  const win = cid.win;
+  const stored = read(win);
+  // See if we have a stored base cid and whether it is still valid
+  // in terms of expiration.
+  if (stored && !isExpired(stored)) {
+    if (shouldUpdateStoredTime(stored)) {
+      // Once per interval we mark the cid as used.
+      store(win, stored.cid);
+    }
+    cid.baseCid_ = stored.cid;
+    return Promise.resolve(stored.cid);
+  }
+  // We need to make a new one.
+  const seed = getEntropy(win);
+  const newVal = cid.sha384Base64_(seed);
+  // Storing the value may require consent. We wait for the respective
+  // promise.
+  persistenceConsent.then(() => {
+    // The initial CID generation is inherently racy. First one that gets
+    // consent wins.
+    const relookup = read(win);
+    if (!relookup) {
+      store(win, newVal);
+    }
+  });
+  return Promise.resolve(newVal);
```

Comment:
Is this the right thing to return a value even if consent is denied? Or maybe we can reject the promise and use it as a signal that the recipient should not perform any action?